### PR TITLE
Add lock exchange test group

### PR DESCRIPTION
**Pull request by scalandr**
This PR adds the lock-exchange test group in Compass. The setup consists of a 2D rectangular tank filled with two fluids with different densities separated by a vertical board in the middle at the x-axis origin. At time t=0, the central board is disengaged and the gravitational adjustment will proceed. This test group has two test cases, one that uses the standard hydrostatic model, and the second one that uses the nonhydrostatic version of the model.  Both test cases have 3 steps:
`initial_state`, which defines the mesh and initial conditions for the model, `forward` which performs time integration of the model, and `visualize`, which plots the density profile obtained with the model (hydrostatic or nonhydrostatic) at t=5s. 

Checklist
* [x] User's Guide has been updated
* [x] Developer's Guide has been updated
* [x] API documentation in the Developer's Guide (`api.rst`) has any new or modified class, method and/or functions listed
* [x] Documentation has been [built locally](https://mpas-dev.github.io/compass/latest/developers_guide/building_docs.html) and changes look as expected
* [x] Document (in a comment titled `Testing` in this PR) any testing that was used to verify the changes

### CODE REVIEW

**Comment by scalandr** [comment]
At the moment, since the nonhydro PR (https://github.com/E3SM-Project/E3SM/pull/5706) is still in progress, please use this branch
https://github.com/scalandr/E3SM/tree/ocean/nonhydro
for testing.

File: docs/users_guide/ocean/test_groups/lock_exchange.rst [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+    L0 = 0.5
+
+    #Reference height
+    a0 = 0.1
+
+    #Time at which plotting the variables
+    plotTime = 5
+
+hydro
+-----------------
```
Comment (xylar) [inline docs/users_guide/ocean/test_groups/lock_exchange.rst:108]:
```suggestion
hydro
-----
```

File: compass/ocean/tests/lock_exchange/initial_state.py [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+                       np.zeros([nCells, nVertLevels]))
+        ds['fEdge'] = (('nEdges', 'nVertLevels',),
+                       np.zeros([nEdges, nVertLevels]))
+        ds['fVertex'] = (('nVertices', 'nVertLevels',),
+                         np.zeros([nVertices, nVertLevels]))
+
+        # surface fields
+        surfaceStress[:] = 0.0
+        atmosphericPressure[:] = 0.0
+        boundaryLayerDepth[:] = 0.0
```
[the remark this answers is not in the document]
Comment (scalandr) [reply compass/ocean/tests/lock_exchange/initial_state.py]: I got rid of these 3 unused fields.

File: compass/ocean/tests/lock_exchange/hydro/__init__.py [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+
+    """
+
+    def __init__(self, test_group):
+        """
+        Create the test case
+
+        Parameters
+        ----------
+        test_group : compass.ocean.tests.lock_exchange.Hydro
```
[the remark this answers is not in the document]
Comment (scalandr) [reply compass/ocean/tests/lock_exchange/hydro/__init__.py]: Change made.

File: compass/ocean/tests/lock_exchange/nonhydro/__init__.py [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+
+    """
+
+    def __init__(self, test_group):
+        """
+        Create the test case
+
+        Parameters
+        ----------
+        test_group : compass.ocean.tests.lock_exchange.Nonhydro
```
[the remark this answers is not in the document]
Comment (scalandr) [reply compass/ocean/tests/lock_exchange/nonhydro/__init__.py]: Change made.

File: compass/ocean/tests/lock_exchange/lock_exchange.cfg [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+#Number of cells in the y-direction
+ny = 4
+
+#Distance from two cell centers
+dc = 0.001
+
+# Options related to the vertical grid
+[vertical_grid]
+
+#Depth of the bottom of the ocean
```
[the remark this answers is not in the document]
Comment (scalandr) [reply compass/ocean/tests/lock_exchange/lock_exchange.cfg]: Units added.

File: compass/ocean/tests/lock_exchange/lock_exchange.cfg [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+grid_type = uniform
+
+# The type of vertical coordinate (e.g. z-level, z-star)
+coord_type = z-star
+
+# Whether to use "partial" or "full", or "None" to not alter the topography
+partial_cell_type = None
+
+# config options for the solitary wave testcase
+[lock_exchange]
```
[the remark this answers is not in the document]
Comment (scalandr) [reply compass/ocean/tests/lock_exchange/lock_exchange.cfg:32]: Units added.

File: compass/ocean/tests/lock_exchange/lock_exchange.cfg [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+# config options for the solitary wave testcase
+[visualize]
+
+#Reference length
+L0 = 0.5
+
+#Reference height
+a0 = 0.1
+
+#Time at which plotting the variables
```
[the remark this answers is not in the document]
Comment (scalandr) [reply compass/ocean/tests/lock_exchange/lock_exchange.cfg]: Units added.

File: docs/users_guide/ocean/test_groups/lock_exchange.rst [hunk 8]
```diff
@@ -0,0 +1,128 @@
...
+tank. At time t=0, the central board is disengaged and the gravitational 
+adjustment will proceed.
+
+The domain is a rectangular tank periodic y with a horizontal length of 
+50 cm and a static water height of 10 cm without the topographic change 
+in the tank. The horizontal and vertical resolution are both 0.001 m.
+Salinity is constant throughout the domain (at 35 PSU) and the initial
+velocity in zero. The initial density is bimodal with the lower density
+of 1023.05 kg m^{-3} on the left side of the domain and the higher
+density of 1026.95 kg m^{-3} on the right side.
```
[the remark this answers is not in the document]
Comment (scalandr) [reply docs/users_guide/ocean/test_groups/lock_exchange.rst]: Change made.

File: compass/ocean/tests/lock_exchange/lock_exchange.cfg [hunk 9]
```diff
@@ -0,0 +1,68 @@
+# config options for the horizontal grid
+[horizontal_grid]
+
+#Number of cells in the x-direction
+nx = 500
+
+#Number of cells in the y-direction
+ny = 4
+
+#Distance from two cell centers
```
[the remark this answers is not in the document]
Comment (scalandr) [reply compass/ocean/tests/lock_exchange/lock_exchange.cfg]: Units added.

File: docs/users_guide/ocean/test_groups/lock_exchange.rst [hunk 10]
```diff
@@ -0,0 +1,128 @@
+.. _ocean_lock_exchange:
+
+lock_exchange
+==================
```
[the remark this answers is not in the document]
Comment (scalandr) [reply docs/users_guide/ocean/test_groups/lock_exchange.rst]: Done.

File: docs/users_guide/ocean/test_groups/lock_exchange.rst (hunk 1 above)
Comment (scalandr) [reply docs/users_guide/ocean/test_groups/lock_exchange.rst]: Done.

File: docs/users_guide/ocean/test_groups/lock_exchange.rst [hunk 11]
```diff
@@ -0,0 +1,128 @@
...
+-----------------
+
+The hydrostatic test case uses the standard hydrostatic model
+to run the lock-exchange test. The density plot shows that 
+the density fronts cannot develop in the upper and lower layer.
+This happens because of the hydrostatic assumption, which 
+prevents the generation of the Kelvin-Helmholtz instability.
+
+nonhydro
+-----------------
```
[the remark this answers is not in the document]
Comment (scalandr) [reply docs/users_guide/ocean/test_groups/lock_exchange.rst]: Done.

**Comment by scalandr** [comment]
I ran both test cases on Chicoma and Anvil with the gnu compiler and the expected results are produced

**hydro**
hydro
**nonhydro**
nonhydro

**Review by xylar — APPROVED** [review]
@scalandr, a beautiful job!  Excited to have this.  I really shows off the nonhydro capability, doesn't it!

**Comment by xylar** [comment]
@vanroekel, if you give this another look and are happy, I'll merge it.

**Review by vanroekel — APPROVED** [review]
I had another look through and this looks great to me.  Approving by visual inspection and testing by others.

Thanks @scalandr!

**Pull request merged**